src/lib.rs

```diff
@@ -151,6 +150,12 @@ pub struct Element {
     /// item insertion order.
     pub attributes: AttributeMap<String, String>,
 
+    /// Attribute namespaces, such as `xml:lang="en"` in the below snippet:
+    /// ```xml
+    /// <mdui:DisplayName xml:lang="en">TestShib Test IdP</mdui:DisplayName>
+    /// ```
+    pub attribute_namespaces: AttributeMap<String, Namespace>,
```

Comment:
I think a better API would be to add a new `Attribute` struct similar to the `Element` struct

```rust
pub struct Attribute {
    pub name: String,
    pub namespace: Option<String>,
    pub prefix Option<String>,
}
```

and have that be stored in `Element::attributes`.

IIRC attributes can also be provided multiple times, so maybe `Element::attributes` should just be a `Vec` of such structs, and the struct should also include the attribute value.

Reply:
I think the underlying xml library will never give us the same attribute/ns combo with the same value twice.

Reply:
Hello, I hope you don't mind me chiming in on this PR but I also have need of this feature. Unfortunately the PR as it currently stands doesn't work for my use case, which is where I have two attributes with the same name but different namespaces, e.g.

`<tag attribute="foo" extended:attribute="bar"/>`

So Sebastian's suggestion seems like a good one, although it sounds like it would break backward compatibility if it changed the current `Element::attributes` member.

Reply:
Uh, does that happen even in the case of:

```xml
<tag a:b="A" b:b="B"/>
```

?

Reply:
It looks that way. Given some XML like this:
`<yv:member type="i" yv:type="[BookingType]" b:type="another type"/>`

only the last 'type' attribute is present in the `Element::attributes` and `Element::attribute_namespaces` members:

`attributes: {"type": "another type"} attribute_namespaces: {"type": Namespace(Namespace({"b": "=http://nowhere.com/schemas/dbus-extensions-v1.0"}))}`

Reply:
Thank you. I thought I'd tested that. That's a big problem. I'll take another look!

Reply:
Okay I see what's wrong. You're right @tombailo that @sdroege made a good suggestion. My not acting on it wasn't a judgment on the quality of the suggestion, my lack of action was just due to the fact the main struggle in this PR is keeping backwards compatibility, which I didn't even manage fully.

I'm going to think some more on the problem and see if I can't find some happy medium between a total API break and my broken implementation. Thanks for testing my PR!

Reply:
You're welcome!

Regarding API compatibility, I can't really see how to see how to fix this without a breaking change. It's an unarguable fact that attribute names can be namespace-qualified and that names in different namespaces need to be discriminated. One option could be to use the namespace-qualified name in `attributes`, but that would be inconsistent with the rest of the `Element` API where names and namespaces are broken out and handled separately. So to me it looks like an API change is inevitable, but I have no idea how big a deal that would be.

Of course ultimately it's up to the maintainers, and I'm pretty new to Rust development, but that's my 2 cents for what it's worth.

Reply:
I'm considering storing both by default with a compile-time flag to disable the duplication. Thoughts?

Reply:
Thanks for the thoughts. I agree with you and will make a semver breaking change. It's up to @eminence whether to pull it or not. :-)